From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Vivek=20Das=C2=A0Mohapatra?= <vivek@collabora.com>
Date: Wed, 16 Apr 2025 15:14:34 +0100
Subject: Revert "xwayland: Restrict allow commit to the window manager"

Steamwebhelper seems to need access to this (previously unrestricted)
functionality, otherwise we get a short black screen interval instead
of a seamless boot display.

This reverts commit 2cc869626a5728d8bd80241322546f98df96094d.
---
 hw/xwayland/xwayland-screen.c | 21 ---------------------
 1 file changed, 21 deletions(-)

diff --git a/hw/xwayland/xwayland-screen.c b/hw/xwayland/xwayland-screen.c
index c83adda9fd60..efe3205b3551 100644
--- a/hw/xwayland/xwayland-screen.c
+++ b/hw/xwayland/xwayland-screen.c
@@ -182,32 +182,11 @@ xwl_property_callback(CallbackListPtr *pcbl, void *closure,
 }
 
 #ifdef XACE
-#define readOnlyPropertyAccessMask (DixReadAccess |\
-                                    DixGetAttrAccess |\
-                                    DixListPropAccess |\
-                                    DixGetPropAccess)
-
 static void
 xwl_access_property_callback(CallbackListPtr *pcbl, void *closure,
                              void *calldata)
 {
-    XacePropertyAccessRec *rec = calldata;
-    PropertyPtr prop = *rec->ppProp;
-    ClientPtr client = rec->client;
-    Mask access_mode = rec->access_mode;
-    ScreenPtr pScreen = closure;
-    struct xwl_screen *xwl_screen = xwl_screen_get(pScreen);
-
-    if (prop->propertyName == xwl_screen->allow_commits_prop) {
-        /* Only the WM and the Xserver itself */
-        if (client != serverClient &&
-            client->index != xwl_screen->wm_client_id &&
-            (access_mode & ~readOnlyPropertyAccessMask) != 0)
-            rec->status = BadAccess;
-    }
 }
-
-#undef readOnlyPropertyAccessMask
 #endif /* XACE */
 
 static void

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Vivek=20Das=C2=A0Mohapatra?= <vivek@collabora.com>
Date: Tue, 22 Apr 2025 17:41:24 +0100
Subject: Revert "xwayland/present: Check allow_commits in xwl_present_flip"

steamwebhelper currently sets _XWAYLAND_ALLOW_COMMITS to (uint32_t)0

This is interpreted as FALSE when xwayland is checking whether commits
are allowed: This was not checked prior to this commit and needs to be
reverted to allow steamwebhelper to invoke xwl_present_flip.

This reverts commit 7c25024879824cc96dc930e1af9e4c203eea740f.
---
 hw/xwayland/xwayland-present.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hw/xwayland/xwayland-present.c b/hw/xwayland/xwayland-present.c
index 7b9a4ac32acd..d29411d2c170 100644
--- a/hw/xwayland/xwayland-present.c
+++ b/hw/xwayland/xwayland-present.c
@@ -870,7 +870,7 @@ xwl_present_flip(present_vblank_ptr vblank, RegionPtr damage)
     struct xwl_present_event    *event = xwl_present_event_from_vblank(vblank);
     Bool                        implicit_sync = TRUE;
 
-    if (!xwl_window || !xwl_window->allow_commits)
+    if (!xwl_window)
         return FALSE;
 
     buffer = xwl_pixmap_get_wl_buffer(pixmap);

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Edmundson <kde@davidedmundson.co.uk>
Date: Wed, 15 Apr 2026 14:37:26 +0100
Subject: libei: Forward text as keysyms where possible

Forwarding keycodes from xtest has multiple problematic parts.

Scripts like `xdotool type y` will look up symbols from the first group
rather than the active keyboard group. It then happens to work on X11
because it forces the current xkb layout to 0 before replaying the text
event. If the keysym it wants to type is not in the current keymap the
xtest client may replace the X11 keymap.

All of these are best solved by determining the keysym on the X11 side
and then sending that to the compositor.
---
 hw/xwayland/xwayland-xtest.c | 40 +++++++++++++++++++++++++++++++-----
 1 file changed, 35 insertions(+), 5 deletions(-)

diff --git a/hw/xwayland/xwayland-xtest.c b/hw/xwayland/xwayland-xtest.c
index c2aacca661c3..54094ab30712 100644
--- a/hw/xwayland/xwayland-xtest.c
+++ b/hw/xwayland/xwayland-xtest.c
@@ -25,6 +25,9 @@
 
 #include <xwayland-config.h>
 
+#include "dixstruct.h"
+#include "xkbsrv.h"
+
 #include <inputstr.h>
 #include <inpututils.h>
 #include <libgen.h>
@@ -72,7 +75,7 @@ struct xwl_ei_client {
     struct xorg_list link;      /* in clients_for_reuse */
     ClientPtr client;           /* can be NULL if the X11 client is gone */
     char *cmdline;
-    bool accept_pointer, accept_keyboard, accept_abs;
+    bool accept_pointer, accept_keyboard, accept_abs, accept_text;
     struct ei *ei;
     int ei_fd;
 #ifdef XWL_HAS_EI_PORTAL
@@ -82,6 +85,7 @@ struct xwl_ei_client {
     struct ei_seat *ei_seat;
     struct ei_device *ei_pointer;
     struct ei_device *ei_keyboard;
+    struct ei_device *ei_text;
     struct xorg_list abs_devices;
     struct xorg_list pending_emulated_events;
 
@@ -177,6 +181,13 @@ add_ei_device(struct xwl_ei_client *xwl_ei_client, struct ei_device *device)
         }
     }
 
+    if (ei_device_has_capability(device, EI_DEVICE_CAP_TEXT) &&
+        xwl_ei_client->ei_text == NULL) {
+        fprintf(stderr, "Xwayland: found ei text device");
+        xwl_ei_client->ei_text = ei_device_ref(device);
+        used = true;
+    }
+
     if (!used)
         ei_device_close(device);
 }
@@ -375,6 +386,7 @@ setup_ei(ClientPtr client)
     xwl_ei_client->accept_pointer = false;
     xwl_ei_client->accept_keyboard = false;
     xwl_ei_client->accept_abs = false;
+    xwl_ei_client->accept_text = false;
     xwl_ei_client->ei = ei;
     xwl_ei_client->ei_fd = -1;
     xwl_ei_client->client = client;
@@ -423,6 +435,8 @@ xwl_ei_start_emulating(struct xwl_ei_client *xwl_ei_client)
     xorg_list_for_each_entry(abs, &xwl_ei_client->abs_devices, link) {
         ei_device_start_emulating(abs->device, sequence);
     }
+    if (xwl_ei_client->ei_text)
+        ei_device_start_emulating(xwl_ei_client->ei_text, sequence);
 }
 
 static void
@@ -533,7 +547,7 @@ xwl_send_abs_event_to_ei(struct xwl_ei_client *xwl_ei_client, int sx, int sy)
 }
 
 static bool
-xwl_send_event_to_ei(struct xwl_ei_client *xwl_ei_client,
+xwl_send_event_to_ei(struct xwl_ei_client *xwl_ei_client, DeviceIntPtr dev,
                      int type, int detail, int flags, const ValuatorMask *mask)
 {
     struct ei *ei = xwl_ei_client->ei;
@@ -592,6 +606,14 @@ xwl_send_event_to_ei(struct xwl_ei_client *xwl_ei_client,
                 return false;
 
             ei_device = xwl_ei_client->ei_keyboard;
+            if (xwl_ei_client->ei_text) {
+                int sym = XkbKeySym(dev->master->key->xkbInfo->desc, detail, 0);
+                if (sym > 0) {
+                    ei_device_text_keysym(xwl_ei_client->ei_text, sym, type == KeyPress);
+                    ei_device_frame(xwl_ei_client->ei_text, ei_now(ei));
+                    break;
+                }
+            }
             ei_device_keyboard_key(ei_device, detail - 8, type == KeyPress);
             ei_device_frame(ei_device, ei_now(ei));
             break;
@@ -689,7 +711,7 @@ xwayland_xtest_send_events(DeviceIntPtr dev,
     }
 
     if (accept) {
-        xwl_send_event_to_ei(xwl_ei_client, type, detail, flags, mask);
+        xwl_send_event_to_ei(xwl_ei_client, dev, type, detail, flags, mask);
     }
     else {
         debug_ei("Not yet connected to EIS, queueing events\n");
@@ -707,6 +729,7 @@ xwl_dequeue_emulated_events(struct xwl_ei_client *xwl_ei_client)
     xorg_list_for_each_entry_safe(xwl_emulated_event, next_xwl_emulated_event,
         &xwl_ei_client->pending_emulated_events, link) {
         sent = xwl_send_event_to_ei(xwl_ei_client,
+                                    xwl_emulated_event->dev,
                                     xwl_emulated_event->type,
                                     xwl_emulated_event->detail,
                                     xwl_emulated_event->flags,
@@ -740,6 +763,9 @@ xwl_ei_update_caps(struct xwl_ei_client *xwl_ei_client,
         if (ei_device == abs->device)
             xwl_ei_client->accept_abs = true;
     }
+    if (ei_device == xwl_ei_client->ei_text) {
+        xwl_ei_client->accept_text = true;
+    }
 }
 
 static bool
@@ -750,7 +776,10 @@ xwl_ei_devices_are_ready(struct xwl_ei_client *xwl_ei_client)
         (xwl_ei_client->accept_pointer ||
          !ei_seat_has_capability(xwl_ei_client->ei_seat, EI_DEVICE_CAP_POINTER)) &&
         (xwl_ei_client->accept_abs ||
-         !ei_seat_has_capability(xwl_ei_client->ei_seat, EI_DEVICE_CAP_POINTER_ABSOLUTE)))
+         !ei_seat_has_capability(xwl_ei_client->ei_seat, EI_DEVICE_CAP_POINTER_ABSOLUTE)) &&
+        (xwl_ei_client->accept_text ||
+         !ei_seat_has_capability(xwl_ei_client->ei_seat, EI_DEVICE_CAP_TEXT))
+        )
         return true;
 
     return false;
@@ -803,7 +832,8 @@ xwl_handle_ei_event(int fd, int ready, void *data)
                                               EI_DEVICE_CAP_POINTER_ABSOLUTE,
                                               EI_DEVICE_CAP_BUTTON,
                                               EI_DEVICE_CAP_SCROLL,
-                                              EI_DEVICE_CAP_KEYBOARD, NULL);
+                                              EI_DEVICE_CAP_KEYBOARD,
+                                              EI_DEVICE_CAP_TEXT, NULL);
                 }
                 break;
             case EI_EVENT_SEAT_REMOVED:
